### PR TITLE
Voyager Backend

### PR DESCRIPTION
**Pull request by sky-2002**
This PR adds [voyager](https://github.com/spotify/voyager) backend
Related issue: #40

### CODE REVIEW

File: vicinity/backends/voyager.py [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    supported_metrics = {Metric.COSINE, Metric.EUCLIDEAN}
+    inverse_metric_mapping = {
+        Metric.COSINE: "cosine",
+        Metric.EUCLIDEAN: "l2",
+    }
+
+    metric_int_mapping = {
+        "l2": 0,
+        "cosine": 2,
+    }
```
Comment (Pringled) [inline vicinity/backends/voyager.py:35]: I think inner_product is also supported in Voyager, I would also add that here as one of the metrics as we also support it (see https://github.com/MinishLab/vicinity/blob/main/vicinity/utils.py#L63).

File: vicinity/backends/voyager.py [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+        """Save the vectors to a path."""
+        path = Path(base_path) / "index.bin"
+        self.index.save(str(path))
+        self.arguments.dump(base_path / "arguments.json")
+
+    def insert(self, vectors: npt.NDArray) -> None:
+        """Insert vectors into the backend."""
+        self.index.add_items(vectors)
+
+    def delete(self, indices: list[int]) -> None:
```
Comment (Pringled) [inline vicinity/backends/voyager.py:102]: At the moment we have decided not to support deletion for ANN backends (see https://github.com/MinishLab/vicinity/blob/main/vicinity/backends/hnsw.py#L101 for example) for the reason you also wrote in the docstring, and instead encourage users to rebuild the index, can you change this to a NotImplementedError like the other backends?

**Review by Pringled — CHANGES_REQUESTED** [review]
Hey @sky-2002 , thanks a lot for making this PR! I think it's almost good to go, just a couple of small comments.

One other comment: could you also add Voyager to the tested backends? It should be as simple as adding Voyager here: https://github.com/MinishLab/vicinity/blob/main/tests/conftest.py#L32, and running make test to see if everything works as expected.

**Comment by Pringled** [comment]
A couple of other comments: 
- could you also add voyager to the requirements file? You can add a voyager dependency like the one here https://github.com/MinishLab/vicinity/blob/main/pyproject.toml#L65, and then also add it to the `all` dependency group
- could you update the documentation to also include voyager? There's 3 parts where it can be included: [supported backends](https://github.com/MinishLab/vicinity/tree/main?tab=readme-ov-file#supported-backends), [backend parameters](https://github.com/MinishLab/vicinity/tree/main?tab=readme-ov-file#backend-parameters), and [installation](https://github.com/MinishLab/vicinity/tree/main?tab=readme-ov-file#backend-parameters).

Thanks in advance!

File: vicinity/backends/voyager.py [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+class VoyagerBackend(AbstractBackend[VoyagerArgs]):
+    argument_class = VoyagerArgs
+    supported_metrics = {Metric.COSINE, Metric.EUCLIDEAN}
+    inverse_metric_mapping = {
+        Metric.COSINE: "cosine",
+        Metric.EUCLIDEAN: "l2",
+    }
+
+    metric_int_mapping = {
```
Comment (stephantul) [inline vicinity/backends/voyager.py:32]: I'd index this directly using the `Metric` itself as key. It doesn't make a lot of sense to first convert things to strings, and can only lead to issues (e.g., someone changing "l2" to "euclidean" in one of the mappings, but not the other.)

File: vicinity/backends/voyager.py [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+    dim: int = 0
+    metric: str = "cosine"
+    ef_construction: int = 200
+    m: int = 16
+
+
+class VoyagerBackend(AbstractBackend[VoyagerArgs]):
+    argument_class = VoyagerArgs
+    supported_metrics = {Metric.COSINE, Metric.EUCLIDEAN}
+    inverse_metric_mapping = {
```
Comment (stephantul) [inline vicinity/backends/voyager.py:27]: I know it doesn't work in terms of typing right now, but you could just use ints here and lose the `metric_int_mapping`

File: vicinity/backends/voyager.py (hunk 3 above)
Comment (stephantul) [reply vicinity/backends/voyager.py:32]: The typing can be fixed by using a `TypeVar` in combination with a Generic argument on the base class. If you don't feel comfortable doing this, let me know, we can add it.

**Review by stephantul — CHANGES_REQUESTED** [review]
Thanks for adding this 🙏 ! I second Thomas his comments about the tests, docs and requirements.

File: vicinity/backends/voyager.py (hunk 1 above)
Comment (Pringled) [reply vicinity/backends/voyager.py:35]: Actually I'm not 100% sure if that is actually a distance metric. Can ignore this for this PR.

Comment (sky-2002) [reply vicinity/backends/voyager.py:35]: okay

File: vicinity/backends/voyager.py [hunk 5]
```diff
@@ -0,0 +1,131 @@
...
+        )
+        index.add_items(vectors)
+        return cls(
+            index,
+            VoyagerArgs(dim=dim, metric=metric, ef_construction=ef_construction, m=m),
+        )
+
+    def query(self, query: npt.NDArray, k: int) -> QueryResult:
+        """Query the backend for the nearest neighbors."""
+        query = normalize(query)
```
[the remark this answers is not in the document]
Comment (sky-2002) [reply vicinity/backends/voyager.py:80]: removed

File: vicinity/backends/voyager.py (hunk 3 above)
Comment (sky-2002) [reply vicinity/backends/voyager.py:32]: I am getting errors(due to `Metric.from_string`) when running `evaluate` method when I use `Metric` as key.

Comment (sky-2002) [reply vicinity/backends/voyager.py:32]: But let me see, will figure out, open to discussions and suggestions

Comment (stephantul) [reply vicinity/backends/voyager.py:32]: Leave it like this, it's fine.

**Comment by stephantul** [comment]
@sky-2002 voyager isn't installed in CI yet, which is why these tests all fail. Could you add it to the list of extra in the Makefile?

**Review by stephantul — APPROVED** [review]

**Review by Pringled — APPROVED** [review]
Approved, thanks for this addition!

**Pull request merged**